velox/exec/AssignUniqueId.cpp

```diff
@@ -60,38 +60,43 @@ RowVectorPtr AssignUniqueId::getOutput() {
   if (input_ == nullptr) {
     return nullptr;
   }
-  generateIdColumn();
+  generateIdColumn(input_->size());
   auto output = fillOutput(input_->size(), nullptr);
   input_ = nullptr;
   return output;
 }
 
-void AssignUniqueId::generateIdColumn() {
-  auto numInput = input_->size();
+void AssignUniqueId::generateIdColumn(vector_size_t size) {
   auto result = results_[0];
   if (!result || !BaseVector::isReusableFlatVector(result)) {
-    result = BaseVector::create(BIGINT(), numInput, operatorCtx_->pool());
+    result = BaseVector::create(BIGINT(), size, operatorCtx_->pool());
     results_[0] = result;
   } else {
-    result->resize(numInput);
+    result->resize(size);
   }
   auto rawResults =
       result->asUnchecked<FlatVector<int64_t>>()->mutableRawValues();
 
   vector_size_t start = 0;
-  while (start < numInput) {
+  while (start < size) {
     if (rowIdCounter_ >= maxRowIdCounterValue_) {
-      rowIdCounter_ = rowIdPool_->fetch_add(kRowIdsPerRequest);
-      maxRowIdCounterValue_ =
-          std::min({rowIdCounter_ + kRowIdsPerRequest, kMaxRowId});
+      requestRowIds();
     }
 
-    auto end = std::min(numInput, (int32_t)(start + kRowIdsPerRequest));
-
+    auto end = std::min(size, (int32_t)(start + kRowIdsPerRequest));
+    VELOX_CHECK_EQ((rowIdCounter_ + end - 1) & uniqueValueMask_, 0);
     std::iota(
         rawResults + start, rawResults + end, uniqueValueMask_ | rowIdCounter_);
     rowIdCounter_ += end;
     start = end;
   }
 }
+
+void AssignUniqueId::requestRowIds() {
+  rowIdCounter_ = rowIdPool_->fetch_add(kRowIdsPerRequest);
+  maxRowIdCounterValue_ =
+      std::min({rowIdCounter_ + kRowIdsPerRequest, kMaxRowId});
+  VELOX_CHECK_LT(
```

Comment:
@majetideepak I am not aware of `VELOX_DCHECK_LT`? What's the different between `VELOX_DCHECK_LT` && `VELOX_CHECK_LT`?

Reply:
Why do we need this check? `std::min({rowIdCounter_ + kRowIdsPerRequest, kMaxRowId});` on the previous line ensures that `maxRowIdCounterValue_ <= kMaxRowId` and it seems valid to have maxRowIdCounterValue_ equal to kMaxRowId, no? Otherwise, are we not allowing the last batch of IDs just before the limit?